### PR TITLE
Performance improvements for the uint256 and uint160 classes

### PR DESCRIPTION
**Pull request by maciejzaleski**

### CODE REVIEW

**Comment by dangershony** [comment]
Do you have benchmark on the improvement?

**Comment by maciejzaleski** [comment]
I don't have the overall performance difference as it would be difficult to quantify. I have prepared some benchmarks for the implemented changes which should give you an idea of the benefits.

**Benchmarks source code**: [Benchmarks.zip](https://github.com/stratisproject/StratisBitcoinFullNode/files/2963792/Benchmarks.zip)

**Statistics**
Benchmarks

**Comment by MithrilMan** [comment]
@maciejzaleski would be interesting to try another updates: applying [MethodImpl(MethodImplOptions.AggressiveInlining)] on top of Utils.ToBytes, utils ToUInt32, and so on
we could have a lot of performance improvement if the compiler didn't inlined them already

**Comment by maciejzaleski** [comment]
I have just run another test purely to verify performance difference of `pn[]` vs `pn0`-`pn7`.
Surprisingly the existing approach of using `pn0`-`pn7` is more optimal.

|                 Method | ArraySize |     Mean |     Error |    StdDev | Ratio | Rank | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|----------------------- |---------- |---------:|----------:|----------:|------:|-----:|------------:|------------:|------------:|--------------------:|
|  CreateUInt256_Current |         1 | 22.43 ns | 0.1407 ns | 0.1316 ns |  1.00 |    1 |      0.0133 |           - |           - |                56 B |
|                        |           |          |           |           |       |      |             |             |             |                     |
| CreateUInt256_Proposed |         1 | 28.19 ns | 0.1765 ns | 0.1565 ns |  1.00 |    1 |      0.0209 |           - |           - |                88 B |

**Comment by maciejzaleski** [comment]
> @maciejzaleski would be interesting to try another updates: applying `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on top of Utils.ToBytes, utils ToUInt32, and so on
> we could have a lot of performance improvement if the compiler didn't inlined them already

I had a look on this but I'm not sure this is a good idea. You need to annotate methods which you want to be inlined and given that inlining will consume more memory and we use uint256 a lot it is not a good idea.

**Comment by MithrilMan** [comment]
> > @maciejzaleski would be interesting to try another updates: applying `[MethodImpl(MethodImplOptions.AggressiveInlining)]` on top of Utils.ToBytes, utils ToUInt32, and so on
> > we could have a lot of performance improvement if the compiler didn't inlined them already
> 
> I had a look on this but I'm not sure this is a good idea. You need to annotate methods which you want to be inlined and given that inlining will consume more memory and we use uint256 a lot it is not a good idea.

What you mean about "consume more memory"?
Inlining should increase only the size of the compiled code, not an issue

File: src/NBitcoin/UInt256.cs [hunk 1]
```diff
@@ -22,7 +22,7 @@ public uint256 Value
                 }
             }
 
-            public uint256 MaxValue => uint256.Parse("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
+            public static uint256 MaxValue => uint256.Parse("ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
```
Comment (NicolasDorier) [inline src/NBitcoin/UInt256.cs]: maybe should just be a singleton

**Comment by NicolasDorier** [comment]
Have you tried with my version of uint256? I am using spans instead of byte array which remove quite a bunch of allocation

**Comment by lontivero** [comment]
I like what you did for removing those ugly `.Reverse().ToArray()`

**Comment by dangershony** [comment]
>Have you tried with my version of uint256? I am using spans instead of byte array which remove quite a bunch of allocation.
Not yet but its definitely coming in one of the next releases.

**Comment by mikedennis** [comment]
Yeah I agree with Nicolas, if we can use stack instead heap for all these hash allocations should have way less memory fragmentation. If your looking for a volunteer to port some of the span code over from NBitcoin, I'm game!

**Comment by dangershony** [comment]
@mikedennis sure the problem is we target netstandard (you'll have to use debug flags)

**Comment by NicolasDorier** [comment]
@dangershony netstandard 2.1 will have span so you can at least prepare it. Or just target both .netcoreapp2.1 and  netstandard and ifdef as I did in NBitcoin.

**Comment by dangershony** [comment]
I suggest we have equivalent implementations of basic primitives (such as uint256 and BitcoinStream) to NBitcoin so we should bring over any improvements and any changes we propose first get them in to NBitcoin. 

@mikedennis in that case lets get over the changes to IBitcoinSerializer and uint256 using compiler directives (as done in NBitcoin) 
@MithrilMan ?

**Comment by mikedennis** [comment]
> I suggest we have equivalent implementations of basic primitives (such as uint256 and BitcoinStream) to NBitcoin so we should bring over any improvements and any changes we propose first get them in to NBitcoin.
> 
> @mikedennis in that case lets get over the changes to IBitcoinSerializer and uint256 using compiler directives (as done in NBitcoin)

Sounds good - I'll start looking into that.

**Comment by MithrilMan** [comment]
Personally I don't like having to pollute the code with directives to target different platforms, we have either to decide if we target dotnetcore or netstandard because we have already lot of project that target directly dotnetcore so this already confuses me about what direction we are going to take.

Unluckily what we need in dotnetcore will be implemented in .netstandard 3.0 and there isn't even an ETA because netstandard release pace is a lot slower than dotnetcore

Moreover .netstandard 3.0 will be used, IIR, from .Net 4.8 forward

We have to decide if we really need to support .netstandard and why, then we need to decide what to do with current solution that has mixed targets.

P.S.
I think some of the project currently require dotnetcore, like smartcontract, for technical reasons but I cannot confirm this, maybe @codingupastorm @rowandh can help about this

**Comment by MithrilMan** [comment]
here an interesting issue about the need for netstandard
https://github.com/dotnet/standard/issues/859#issuecomment-435348206

**Comment by dangershony** [comment]
🤔  Unless stated otherwise by the architecture the sensible way is to target netstandard for now and add some compilation directives (temporary). But I do see the quick win of just targeting dotnetcore for now (specially that we don't plan to target different platforms in the near future), swapping to netstandard should in theory just be a change to the compilation target and that's it.

**Comment by mikedennis** [comment]
My preference is to do the multi-targeting/conditional compiles (like NBitcoin) for now but once we can upgrade to .net standard 2.1 then clean all that out. I don't think it should be too long and we can try to keep it to serialized and uint256 classes for now.

I thought they would have had 2.1 out already but seems to be taking a while.

**Comment by NicolasDorier** [comment]
NBitcoin keep compatibility up to .NETStandard 1.1, so my pragma are not soon removed :p

**Comment by dangershony** [comment]
>NBitcoin keep compatibility up to .NETStandard 1.1, so my pragma are not soon removed :p  

this is awesome, fortunately we don't have that limitation.

**Comment by mikedennis** [comment]
@dangershony @MithrilMan Working on implementing span will track that work in #1871

**Comment by MithrilMan** [comment]
@mikedennis we agreed internally to move all our projects to dotnetcore2.2, so if you don't want to waste time with directives wait few days or target already the dotnetcore2.2

**Comment by mikedennis** [comment]
@MithrilMan Ok cool that will make this much easier! I've taken a quick look at know where the touch points are, so should be pretty easy once we are running netcoreapp2.2

**Comment by fassadlr** [comment]
@mikedennis I've updated Jeremy's .net core 2.2 upgrade and it is pretty much ready to be merged. We are just awaiting side chain work to be finished and as soon as it is we will merge the upgrade PR and then we can implement these changes. Should be next week!

**Comment by fassadlr** [comment]
Closing as we will raise an internal issue to implement spans once we upgrade to .net Core 2.2.

**Pull request closed without merging**